app/src/main/java/com/concordium/wallet/data/preferences/AuthPreferences.kt

```diff
@@ -142,17 +146,90 @@ class AuthPreferences(val context: Context) :
         setString(SEED_PHRASE, "")
     }
 
-    fun setSeedPhrase(value: String) {
+    suspend fun tryToSetEncryptedSeedPhrase(value: String, password: String): Boolean {
         val seed = Mnemonics.MnemonicCode(value).toSeed()
         val seedEncoded = seed.toHex()
-        setString(SEED_PHRASE, seedEncoded)
+        val encryptedSeed = App.appCore.getCurrentAuthenticationManager().encryptInBackground(password, seedEncoded)
+            ?: return false
+       return setStringWithResult(SEED_PHRASE_ENCRYPTED, encryptedSeed)
     }
 
-    fun getSeedPhrase(): String {
-        return getString(SEED_PHRASE, "")
+    suspend fun getSeedPhrase(password: String): String {
+
+        val seedEncrypted = getString(SEED_PHRASE_ENCRYPTED)
+        val seedUnencrypted = getString(SEED_PHRASE)
+
+        if(seedEncrypted != null && seedUnencrypted == null){
+            //Seed is encrypted and the old seed is deleted, decrypt and return the result
+            return App.appCore.getCurrentAuthenticationManager().decryptInBackground(password, seedEncrypted)
+                ?: return ""
+        }
+
+        if(seedEncrypted != null && seedUnencrypted != null){
+            //Booth encrypted and decrypted seeds present, delete the old seed
+            setString(SEED_PHRASE, null)
+            return seedUnencrypted
+        }
+
+        if(seedUnencrypted != null){
+            //Only the unencrypted seed is present, encrypt it, save it and delete the old seed
+            val encryptedSeed = App.appCore.getCurrentAuthenticationManager().encryptInBackground(password, seedUnencrypted) ?: return ""
+            setStringWithResult(SEED_PHRASE_ENCRYPTED, encryptedSeed).let { saveSuccess ->
+                if (saveSuccess) {
+                        Log.d("ENCRYPTED SEED UPDATED")
+                        setString(SEED_PHRASE, null)
+                        return seedUnencrypted
+                    }
+                }
+        }
+        return ""
+    }
+
+    suspend fun getSeedPhraseDecrypted(decryptKey: SecretKey): String? {
+
+        getString(SEED_PHRASE_ENCRYPTED)?.let {seedEncrypted ->
+            return App.appCore.getOriginalAuthenticationManager()
+                .decryptInBackground(decryptKey, seedEncrypted)
+        }
+        return getString(SEED_PHRASE)
+    }
+
+    suspend fun checkAndTryToEncryptSeed(password: String): Boolean {
+
+        val seedEncrypted = getString(SEED_PHRASE_ENCRYPTED)
+        val seedUnencrypted = getString(SEED_PHRASE)
+
+        if(seedEncrypted != null && seedUnencrypted == null){
+            //Encrypted seed is stored, unencrypted is erased
+            return true
+        }
+
+        if(seedEncrypted != null && seedUnencrypted != null){
+            //Booth encrypted and decrypted seeds present, delete the old seed
+            return setStringWithResult(SEED_PHRASE, null)
```

Comment:
We have 3 points where we access the seed.

- from the passcode screen, we check to see if the seed is encrypted and the old seed is deleted. if its not delete it(this is the main check).
- when we change the password, the old seed needs to be re-encrypted with the new password. Here we also attempt to delete the old seed
- and when we need to read the seed and use it. Here we also try to delete the seed
 
In any of these 3 cases the old seed could be present and we try to delete it. I could remove the other instances and leave only the logic on the passcode screen. This will simplify things but there is a case where the user can login in the app using the passcode, if the encryption fails or the delete old seed fails, the user can use the app as normal even use the old seed and then leaves the app, the seed is left unencrypted on the device until he opens the app again and uses the passcode(we don't know when this will happen). The other checks and attempts only improve our chances that the seed is encrypted and the old deleted.